spring-cloud-function-adapters/spring-cloud-function-adapter-google-cloud-function/src/main/java/org/springframework/cloud/function/adapter/gcloud/GcloudSpringBootHttpRequestHandler.java

```diff
@@ -0,0 +1,196 @@
+			return toOtionalIfEmpty(requestBody);
+		}
+
+	}
+
+	private Object toOtionalIfEmpty(String requestBody) {
+		return requestBody.isEmpty() ? Optional.empty() : requestBody;
+	}
+
+	private MessageHeaders getHeaders(HttpRequest event, String requestBody) {
+		Map<String, Object> headers = new HashMap<String, Object>();
+
+		if (event.getHeaders() != null) {
+			headers.putAll(event.getHeaders());
+		}
+		if (event.getQueryParameters() != null) {
+			headers.putAll(event.getQueryParameters());
+		}
+		if (event.getUri() != null) {
+			headers.put("path", event.getPath());
+		}
+
+		if (event.getMethod() != null) {
+			headers.put("httpMethod", event.getMethod());
+		}
+
+		headers.put("request", requestBody);
+		return new MessageHeaders(headers);
+	}
+
+	protected boolean functionAcceptsMessage() {
+		return this.getInspector().isMessage(function());
+	}
+
+	@SuppressWarnings("unchecked")
+	protected <T> T result(Object input, Publisher<?> output, HttpResponse resp) {
+		List<T> result = new ArrayList<>();
+		for (Object value : Flux.from(output).toIterable()) {
+			result.add((T) convertOutput1(value, resp));
+		}
+		if (isSingleValue(input) && result.size() == 1) {
+			return result.get(0);
+		}
+		return (T) result;
+	}
+
+
+	private boolean isSingleValue(Object input) {
+		return !(input instanceof Collection);
+	}
+
+	private Flux<?> extract(Object input) {
+		if (input instanceof Collection) {
+			return Flux.fromIterable((Iterable<?>) input);
+		}
+		return Flux.just(input);
+	}
+
+	public void service(HttpRequest httpRequest, HttpResponse httpResponse) throws Exception {
+		Thread.currentThread()
+			.setContextClassLoader(GcloudSpringBootHttpRequestHandler.class.getClassLoader());
+		initialize(new TestExecutionContext());
+
+		Publisher<?> output = apply(extract(convert(httpRequest)));
+		BufferedWriter writer = httpResponse.getWriter();
+		Object result = result(httpRequest, output, httpResponse);
+		if (returnsOutput()) {
+			writer.write(gson.toJson(result));
+			writer.flush();
+		}
+		httpResponse.setStatusCode(200);
+	}
+
+	protected O convertOutput1(Object output, HttpResponse resp) {
+		if (functionReturnsMessage(output)) {
+			Message<?> message = (Message<?>) output;
+			for (Map.Entry<String, Object> entry : message.getHeaders().entrySet()) {
+				Object values = entry.getValue();
+				if (values instanceof List) {
+					for (Object value : (List) values) {
+						if (value != null) {
+							resp.appendHeader(entry.getKey(), value.toString());
+						}
+					}
+				}
+				else if (values != null) {
+					resp.appendHeader(entry.getKey(), values.toString());
+				}
+			}
+			return (O) message.getPayload();
+		}
+		else {
+			return (O) output;
+		}
+	}
+
+	boolean returnsOutput() {
+		return !this.getInspector().getOutputType(function()).equals(Void.class);
+	}
+
+	protected boolean functionReturnsMessage(Object output) {
+		return output instanceof Message;
+	}
+}
+
+class TestExecutionContext implements Context {
```

Comment:
Make this into a `private static` nested class inside the adapter; I think it's best practice to keep only 1 top-level class per .java file.